File: src/class034/C6_SortList.cpp
#include <iostream>
using namespace std;

// 链表节点类
struct ListNode {
    int val;
    ListNode* next;
    ListNode(int x) : val(x), next(nullptr) {}
};

ListNode* start = nullptr;  // 合并操作后的链表头
ListNode* end = nullptr;    // 合并操作后的链表尾

ListNode* findEnd(ListNode* s, int k) {
    // 返回链表段s的k个节点之后的节点或链表尾
    while (s != nullptr && s->next != nullptr && --k != 0) {
        s = s->next;
    }
    return s;
}

void merge(ListNode* l1, ListNode* r1, ListNode* l2, ListNode* r2) {
    // 合并两个已排序的链表段，并设置start为合并后链表的头，end为尾
    ListNode* pre = nullptr;

    if (l1->val <= l2->val) {
        start = l1;
        pre = l1;
        l1 = l1->next;
    } else {
        start = l2;
        pre = l2;
        l2 = l2->next;
    }

    while (l1 != nullptr && l2 != nullptr) {
        if (l1->val <= l2->val) {
            pre->next = l1;
            pre = l1;
            l1 = l1->next;
        } else {
            pre->next = l2;
            pre = l2;
            l2 = l2->next;
        }
    }

    if (l1 != nullptr) {
        pre->next = l1;
        end = r1;
    } else {
        pre->next = l2;
        end = r2;
    }
}

ListNode* sortList(ListNode* head) {
    // 计算链表长度
    int n = 0;
    ListNode* cur = head;
    while (cur != nullptr) {
        n++;
        cur = cur->next;
    }

    for (int step = 1; step < n; step <<= 1) {
        // 第一组独立处理
        ListNode* l1 = head;
        ListNode* r1 = findEnd(l1, step);
        ListNode* l2 = r1 != nullptr ? r1->next : nullptr;
        ListNode* r2 = findEnd(l2, step) if l2 != nullptr : nullptr;
        ListNode* next = r2 != nullptr ? r2->next : nullptr;

        if (r1 != nullptr) r1->next = nullptr;
        if (r2 != nullptr) r2->next = nullptr;

        merge(l1, r1, l2, r2);

        head = start;
        ListNode* lastTeamEnd = end;

        while (next != nullptr) {
            l1 = next;
            r1 = findEnd(l1, step);
            l2 = r1 != nullptr ? r1->next : nullptr;

            if (l2 == nullptr) {
                lastTeamEnd->next = l1;
                break;
            }

            r2 = findEnd(l2, step);
            next = r2 != nullptr ? r2->next : nullptr;

            if (r1 != nullptr) r1->next = nullptr;
            if (r2 != nullptr) r2->next = nullptr;

            merge(l1, r1, l2, r2);

            lastTeamEnd->next = start;
            lastTeamEnd = end;
        }
    }

    return head;
}

File: src/class034/C6_SortList.py
# 排序链表
# 要求时间复杂度O(n*logn)，额外空间复杂度O(1)，还要求稳定性

class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

def sort_list(head):
    # 计算链表长度
    def get_length(node):
        length = 0
        while node:
            length += 1
            node = node.next
        return length

    # 找到链表中某段的尾部，返回起点s往下数k个节点
    def find_end(s, k):
        while s and s.next and k > 1:
            s = s.next
            k -= 1
        return s

    # 合并两段有序链表
    def merge(l1, r1, l2, r2):
        if not l1:
            return l2, r2
        if not l2:
            return l1, r1

        start = end = None
        if l1.val <= l2.val:
            start = end = l1
            l1 = l1.next
        else:
            start = end = l2
            l2 = l2.next

        while l1 and l2:
            if l1.val <= l2.val:
                end.next = l1
                end = l1
                l1 = l1.next
            else:
                end.next = l2
                end = l2
                l2 = l2.next

        if l1:
            end.next = l1
            end = r1
        else:
            end.next = l2
            end = r2

        return start, end

    n = get_length(head)

    for step in range(1, n, step := step * 2):
        l1 = head
        last_team_end = None

        while l1:
            r1 = find_end(l1, step)
            l2 = r1.next if r1 else None
            if not l2:
                if last_team_end:
                    last_team_end.next = l1
                break

            r2 = find_end(l2, step)
            next = r2.next if r2 else None

            r1.next = None
            r2.next = None

            merged_start, merged_end = merge(l1, r1, l2, r2)
            
            if last_team_end:
                last_team_end.next = merged_start
            else:
                head = merged_start

            last_team_end = merged_end

            l1 = next

    return head
